src/ucp/core/ucp_rkey.c

```diff
@@ -213,61 +222,323 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
+{
+    ucp_context_h context = memh->context;
+
+    if (flags & UCP_MEMH_PACK_FLAG_EXPORT) {
+        ucs_assert(!rkey_compat);
+        return ucp_memh_exported_packed_size(
+                context, context->export_md_map[memh->mem_type]);
+    }
+
+    if (rkey_compat) {
+        return ucp_rkey_packed_size(context, memh->md_map,
+                                    UCS_SYS_DEVICE_ID_UNKNOWN, 0);
+    }
+
+    ucs_fatal("packing rkey using ucp_memh_pack() is unsupported");
+}
+
+static ssize_t ucp_memh_do_pack(ucp_mem_h memh, uint64_t flags,
+                                int rkey_compat, void *memh_buffer)
+{
+    ucp_memory_info_t mem_info;
+
+    if (flags & UCP_MEMH_PACK_FLAG_EXPORT) {
+        return ucp_memh_exported_pack(memh, memh_buffer);
+    }
+
+    if (rkey_compat) {
+        mem_info.type    = memh->mem_type;
+        mem_info.sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+        return ucp_rkey_pack_memh(memh->context, memh->md_map, memh, &mem_info,
+                                  0, NULL, memh_buffer);
+    }
+
+    ucs_fatal("packing rkey using ucp_memh_pack() is unsupported");
+}
+
 static ucs_status_t
 ucp_memh_pack_internal(ucp_mem_h memh, const ucp_memh_pack_params_t *params,
                        int rkey_compat, void **buffer_p, size_t *buffer_size_p)
 {
     ucp_context_h context = memh->context;
-    ucp_memory_info_t mem_info;
     ucs_status_t status;
     ssize_t packed_size;
-    void *rkey_buffer;
+    void *memh_buffer;
     size_t size;
+    uint64_t flags;
 
-    ucs_trace("packing rkeys for buffer %p memh %p md_map 0x%"PRIx64,
-              ucp_memh_address(memh), memh, memh->md_map);
-
-    if (!rkey_compat) {
-        return UCS_ERR_UNSUPPORTED;
-    }
+    ucs_trace("packing memh %p for buffer %p md_map 0x%" PRIx64
+              " export_md_map 0x%" PRIx64, memh, ucp_memh_address(memh),
+              memh->md_map, context->export_md_map[memh->mem_type]);
 
     if (ucp_memh_is_zero_length(memh)) {
         /* Dummy memh, return dummy key */
-        *buffer_p      = &ucp_mem_dummy_buffer;
-        *buffer_size_p = sizeof(ucp_mem_dummy_buffer);
+        if (rkey_compat) {
+            *buffer_p      = &ucp_memh_rkey_dummy_buffer;
+            *buffer_size_p = sizeof(ucp_memh_rkey_dummy_buffer);
+        } else {
+            *buffer_p      = &ucp_memh_dummy_buffer;
+            *buffer_size_p = sizeof(ucp_memh_dummy_buffer);
+        }
         return UCS_OK;
     }
 
     UCP_THREAD_CS_ENTER(&context->mt_lock);
 
-    size        = ucp_rkey_packed_size(context, memh->md_map,
-                                       UCS_SYS_DEVICE_ID_UNKNOWN, 0);
-    rkey_buffer = ucs_malloc(size, "ucp_rkey_buffer");
-    if (rkey_buffer == NULL) {
-        status = UCS_ERR_NO_MEMORY;
+    flags = UCP_PARAM_VALUE(MEMH_PACK, params, flags, FLAGS, 0);
+    size  = ucp_memh_packed_size(memh, flags, rkey_compat);
+
+    if ((flags & UCP_MEMH_PACK_FLAG_EXPORT) &&
+        (context->export_md_map[memh->mem_type] == 0)) {
```

Comment:
please print an error here

Reply:
better to print `ucs_diag`, because user can handle it and fallback from Exported-Offload scheme to basic offload scheme